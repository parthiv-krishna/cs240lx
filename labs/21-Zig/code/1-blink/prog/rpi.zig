extern fn nop() void;

// countdown 'ticks' cycles
pub fn delay(ticks: u32) void {
    var cnt = @as(u32, ticks);
    while (cnt > 0) {
        nop();
        cnt -= @as(u32, 1);
    }
}


