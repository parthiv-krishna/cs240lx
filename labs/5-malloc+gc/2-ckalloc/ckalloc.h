// weak test of malloc/free: checks that subsequent allocations are
// <sizeof (void *)> bytes from each other.

#ifndef __CKALLOC_H__
#define __CKALLOC_H__

#include "src-loc.h"

typedef enum {  ALLOCED = 11, FREED } state_t;

// pull the remainder into the second redzone.
typedef struct ck_hdr {
    struct ck_hdr *next;
    uint32_t nbytes_alloc;  // how much the user requested to allocate.
    uint32_t state;          // state of the block: { ALLOCED, FREED }

    src_loc_t alloc_loc;    // location they called ckalloc() at.

    // used for gc: i didn't cksum these.
    uint32_t refs_start;    // number of pointers to the start of the block.
    uint32_t refs_middle;   // number of pointers to the middle of the block.

    uint16_t mark;          // 0 initialize.
} hdr_t;

// returns pointer to the first header block.
hdr_t *ck_first_hdr(void);

// returns pointer to next hdr or 0 if none.
hdr_t *ck_next_hdr(hdr_t *p);

// get the number of bytes for hdr <h>
static inline unsigned ck_nbytes(hdr_t *h) {
    return h->nbytes_alloc;
}
// get the start of allocated data for <h>
void *ck_hdr_start(hdr_t *h);
// get the end of allocated data for <h>
void *ck_hdr_end(hdr_t *h);

// is <ptr> in block <h>?
unsigned 
ck_ptr_in_block(hdr_t *h, void *ptr);

// is pointer <p> on the allocated list?
int ck_ptr_is_alloced(void *ptr);

#define ckalloc(_n) (ckalloc)(_n, SRC_LOC_MK())
#define ckfree(_ptr) (ckfree)(_ptr, SRC_LOC_MK())
void *(ckalloc)(uint32_t nbytes, src_loc_t loc);
void (ckfree)(void *addr, src_loc_t loc);

#endif
