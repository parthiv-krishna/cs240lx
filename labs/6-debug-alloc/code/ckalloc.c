// implement a simple ckalloc/free that adds ckhdr_t to the 
// allocation.

#include "test-interface.h"
#include "ckalloc.h"

// keep a list of allocated blocks.
static hdr_t *alloc_list;

// returns pointer to the first header block.
hdr_t *ck_first_hdr(void) {
    return alloc_list;
}

// returns pointer to next hdr or 0 if none.
hdr_t *ck_next_hdr(hdr_t *p) {
    if(p)
        return p->next;
    return 0;
}


void *ck_hdr_start(hdr_t *h) {
    return &h[1];
}

// one past the last byte of allocated memory.
void *ck_hdr_end(hdr_t *h) {
    return (char *)ck_hdr_start(h) + ck_nbytes(h);
}

// is ptr in <h>?
unsigned ck_ptr_in_block(hdr_t *h, void *ptr) {
    char *start = (char *)(h + 1);
    char *end = start + h->nbytes_alloc;
    char *p = (char *)ptr;
    return (p >= start && p < end);
}


hdr_t *ck_ptr_is_alloced(void *ptr) {
    for(hdr_t *h = ck_first_hdr(); h; h = ck_next_hdr(h))
        if(ck_ptr_in_block(h,ptr)) {
            // output("found %p in %p\n", ptr, h);
            return h;
        }
    return NULL;
}

static void list_remove(hdr_t **l, hdr_t *h) {
    assert(l);
    hdr_t *prev = *l;
 
    if(prev == h) {
        *l = h->next;
        return;
    }

    hdr_t *p;
    while((p = ck_next_hdr(prev))) {
        if(p == h) {
            prev->next = p->next;
            return;
        }
        prev = p;
    }
    panic("did not find %p in list\n", h);
}

// free a block allocated with <ckalloc>
void (ckfree)(void *addr, src_loc_t l) {
    hdr_t *h = (void *)addr;
    h -= 1;

    if(h->state != ALLOCED)
        loc_panic(l, "freeing unallocated memory: state=%d\n", h->state);
    loc_debug(l, "freeing %p\n", addr);
    
    h->state = FREED;
    memset(h + 1, 0, h->nbytes_alloc);
    assert(ck_ptr_is_alloced(addr));
    kr_free(h);
    list_remove(&alloc_list, h);
    assert(!ck_ptr_is_alloced(addr));
}

// interpose on kr_malloc allocations and
//  1. allocate enough space for a header and fill it in.
//  2. add the allocated block to  the allocated list.
void *(ckalloc)(uint32_t nbytes, src_loc_t loc) {

    hdr_t *h = kr_malloc(nbytes + sizeof *h);

    memset(h, 0, sizeof *h);
    h->nbytes_alloc = nbytes;
    h->state = ALLOCED;
    h->alloc_loc = loc;


    loc_debug(loc, "allocating %p\n", h);

    void *addr = (void *)(h + 1);

    assert(!ck_ptr_is_alloced(addr));
    h->next = alloc_list;
    alloc_list = h;

    assert(ck_ptr_is_alloced(addr));

    return addr;
}
