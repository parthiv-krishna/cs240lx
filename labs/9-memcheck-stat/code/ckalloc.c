// implement a simple ckalloc/free that adds ckhdr_t to the 
// allocation.

#include "test-interface.h"
#include "ckalloc.h"

// keep a list of allocated blocks.
static hdr_t *alloc_list;
static hdr_t *free_list;

unsigned ck_verbose_p = 0;

// returns pointer to the first header block.
hdr_t *ck_first_hdr(void) {
    return alloc_list;
}

// returns pointer to next hdr or 0 if none.
hdr_t *ck_next_hdr(hdr_t *p) {
    if(p)
        return p->next;
    return 0;
}


void *ck_hdr_start(hdr_t *h) {
    return &h[1];
}

// one past the last byte of allocated memory.
void *ck_hdr_end(hdr_t *h) {
    return (char *)ck_hdr_start(h) + ck_nbytes(h);
}

// is ptr in <h>?
unsigned ck_ptr_in_block(hdr_t *h, void *ptr) {
    char *start = (char *)(h + 1);
    char *end = start + h->nbytes_alloc;
    char *p = (char *)ptr;
    return (p >= start && p < end);
}


hdr_t *ck_ptr_is_alloced(void *ptr) {
    for(hdr_t *h = ck_first_hdr(); h; h = ck_next_hdr(h))
        if(ck_ptr_in_block(h,ptr)) {
            // output("found %p in %p\n", ptr, h);
            return h;
        }
    return NULL;
}

static void list_remove(hdr_t **l, hdr_t *h) {
    assert(l);
    hdr_t *prev = *l;
 
    if(prev == h) {
        *l = h->next;
        return;
    }

    hdr_t *p;
    while((p = ck_next_hdr(prev))) {
        if(p == h) {
            prev->next = p->next;
            return;
        }
        prev = p;
    }
    panic("did not find %p in list\n", h);
}


void make_redzone(char *p, unsigned n) {
    memset(p, REDZONE_VAL, n);
}

int check_redzone(char *p, unsigned n) {
    for(unsigned i = 0; i < n; i++) {
        if(p[i] != REDZONE_VAL) {
            printk("found error at %p, %u, %d\n", p, i, p[i]);
            return 4*i + 1;
        }
    }
    return 0;
}

static int ckalloc_count = 1;

// free a block allocated with <ckalloc>
void (ckfree)(void *addr, src_loc_t l) {
    hdr_t *h = (void *)addr;
    h -= 1;

    int ofst = check_redzone(h->rz1, REDZONE_NBYTES);
    if (ofst != 0) {
        trace("ERROR:block %u  corrupted at offset %d\n",
              h->block_id, ofst - 1);
        trace("        logical block id=%u,  nbytes=%u\n", h->block_id, h->nbytes_alloc);
        trace("        Block allocated at: %s:%s:%d\n", h->alloc_loc.file, h->alloc_loc.func, h->alloc_loc.lineno);
        
        panic("corrupted block %u [addr=%p]", h->block_id, addr);
    }        
    
    ofst = check_redzone((char *)addr + h->nbytes_alloc, REDZONE_NBYTES);
    if (ofst != 0) {
        trace("ERROR:block %u  corrupted at offset %d\n",
              h->block_id, ofst - 1 + h->nbytes_alloc);
        trace("        logical block id=%u,  nbytes=%u\n", h->block_id, h->nbytes_alloc);
        trace("        Block allocated at: %s:%s:%d\n", h->alloc_loc.file, h->alloc_loc.func, h->alloc_loc.lineno);
        
        panic("corrupted block %u [addr=%p]", h->block_id, addr);
    }        

    

    if(h->state != ALLOCED)
        loc_panic(l, "freeing unallocated memory: state=%d\n", h->state);
    loc_debug(l, "freeing block=%u [addr=%p]\n", h->block_id, addr);
    
    h->state = FREED;
    make_redzone((char *)(h + 1), h->nbytes_alloc); // nuke data
    assert(ck_ptr_is_alloced(addr));

    list_remove(&alloc_list, h);
    h->next = free_list;
    free_list = h;


    // list_remove(&alloc_list, h);
    assert(!ck_ptr_is_alloced(addr));
}


// interpose on kr_malloc allocations and
//  1. allocate enough space for a header and fill it in.
//  2. add the allocated block to  the allocated list.
void *(ckalloc)(uint32_t nbytes, src_loc_t loc) {


    hdr_t *h = kr_malloc(nbytes + sizeof *h + REDZONE_NBYTES);

    memset(h, 0, sizeof *h);
    h->nbytes_alloc = nbytes;
    h->state = ALLOCED;
    h->alloc_loc = loc;
    h->block_id = ckalloc_count++;

    loc_debug(loc, "allocating block=%u [addr=%p]\n", h->block_id, h);

    void *addr = (void *)(h + 1);

    make_redzone(h->rz1, REDZONE_NBYTES);
    make_redzone((char *)addr + nbytes, REDZONE_NBYTES);

    assert(!ck_ptr_is_alloced(addr));
    h->next = alloc_list;
    alloc_list = h;

    assert(ck_ptr_is_alloced(addr));

    return addr;
}


int ck_heap_errors() {
    int errors = 0;
    int nblocks = 0;
    for(hdr_t *h = alloc_list; h; h = h->next) {
        nblocks++;

        if (h->state == ALLOCED) {
            if (ck_verbose_p) printk("alloc block block=%u [addr=[%p]\n", h->block_id, h);
            if (check_redzone(h->rz1, REDZONE_NBYTES) ||
                check_redzone((char *)(h + 1) + h->nbytes_alloc, REDZONE_NBYTES)) {
                    errors++;
            }
        } else {
            panic("free block %u [addr=%p] in alloc list", h->block_id, h);
        }
    }

    for(hdr_t *h = free_list; h; h = h->next) {
        nblocks++;

        if (h->state == FREED) {
            if (ck_verbose_p) printk("free block block=%u [addr=%p]\n", h->block_id, h);
            if (check_redzone(h->rz1, REDZONE_NBYTES) ||
                check_redzone((char *)(h + 1), h->nbytes_alloc) ||
                check_redzone((char *)(h + 1) + h->nbytes_alloc, REDZONE_NBYTES)) {
                    errors++;
            }
        } else {
            panic("alloc block %u [addr=%p] in free list", h->block_id, h);
        }

    }

    return errors;
}
