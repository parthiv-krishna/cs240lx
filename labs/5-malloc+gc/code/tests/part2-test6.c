// written by jimpo, 5/19/20
#include "rpi.h"
#include "ckalloc.h"

char *test1(void) {
    // allocate then move pointer to middle of allocated region.
    char *p = ckalloc(4);
    trace("alloc returned %p\n", p);
    memset(p, 0, 4);
    return p + 1;
}

void notmain(void) {
    printk("GC test6\n");

    // start heap allocating after the first mb.   give it 1mb to use.
    char *p = test1();
    // print something so compiler is forced to save p.
    trace("test1 returned %p\n", p);

    // this will not warm if you point to the center.
    if(ck_find_leaks(0)) {
        panic("should have no definite leaks!\n");
    }

    if(ck_find_leaks(1)) {
        trace("SUCCESS: detected a maybe leak!\n");
    } else {
        panic("should have found a maybe leak!\n");
    }
}
