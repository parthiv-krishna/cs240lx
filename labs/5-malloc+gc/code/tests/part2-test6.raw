find-ttyusb.c:find_ttyusb:55:FOUND: </dev/ttyUSB0>
opened tty port </dev/ttyUSB0>.
pi-install: tty-usb=</dev/ttyUSB0> program=<./tests/part2-test6.bin> about to boot
HASH: simple_boot: sending 10920 bytes, crc32=be54eb8
waiting for a start
putting code
bootloader: Done.
listening on ttyusb=</dev/ttyUSB0>
GC test6
tests/part2-test6.c:test1:7:allocating 0x17fffc
found 0x17fffc in 0x17ffd8
tests/part2-test6.c:test1:7:successful alloc of 0x17fffc
TRACE:alloc returned 0x17fffc
TRACE:test1 returned 0x17fffd
ck-gc.c:mark_all:204:stack has 26 words
found 0x17fffd in 0x17ffd8
ck-gc.c:mark:97:CK_DEBUG:checking pointer 0x17fffd
ck-gc.c:mark:105:CK_DEBUG:addr 0x7ffffe8: referenced middle = 0x17fffc
---------------------------------------------------------
checking for leaks:
ck-gc.c:sweep_leak:142:CK_DEBUG:checking pointer: 0x17fffc, start=0, middle=1
TRACE:	GC:Checked 1 blocks.
TRACE:		GC:SUCCESS: No leaks found!
----------------------------------------------------------
ck-gc.c:mark_all:204:stack has 26 words
found 0x17fffd in 0x17ffd8
ck-gc.c:mark:97:CK_DEBUG:checking pointer 0x17fffd
ck-gc.c:mark:105:CK_DEBUG:addr 0x7ffffe8: referenced middle = 0x17fffc
---------------------------------------------------------
checking for leaks:
ck-gc.c:sweep_leak:142:CK_DEBUG:checking pointer: 0x17fffc, start=0, middle=1
TRACE:ERROR:GC:MAYBE LEAK of 0x17fffc (no pointer to the start)
TRACE:	nbytes=4
TRACE:	Block allocated at: tests/part2-test6.c:test1:7
TRACE:	GC:Checked 1 blocks.
TRACE:		GC:ERRORS: 0 errors, 1 maybe_errors
----------------------------------------------------------
TRACE:SUCCESS: detected a maybe leak!
DONE!!!

Saw done

bootloader: pi exited.  cleaning up
