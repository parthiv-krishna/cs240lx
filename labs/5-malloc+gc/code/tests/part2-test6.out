find-ttyusb.c:find_ttyusb_first:106:FOUND: </dev/ttyUSB0>
opened tty port </dev/ttyUSB0>.
my-install: tty-usb=</dev/ttyUSB0> program=<./tests/part2-test6.bin> about to boot
******************sending 11492 bytes
waiting for a start
got a start
putting code
put code and got back
bootloader: Done.
listening on ttyusb=</dev/ttyUSB0>
GC test6
sizeof hdr=56, redzone=128
TRACE:allocating 4 bytes
ckalloc.c:ckalloc:170:CK_DEBUG:hdr=0x100000, alloc=0x1000b8, cksum=0x84104c2, (get=0x84104c2)
ckalloc.c:ckalloc:172:CK_DEBUG:header checked
TRACE:ckalloc:allocated 4 bytes, (total=320), ptr=0x1000b8
TRACE:alloc returned 0x1000b8
TRACE:test1 returned 0x1000b9
TRACE:going to check heap: 320 bytes allocated, 1048256 bytes left
TRACE:SUCCESS: checked 1 blocks, detected no errors
TRACE:SUCCESS heap checked out
ck-gc.c:mark_all:207:stack has 30 words
ck-gc.c:mark:94:CK_DEBUG:checking pointer 0x1000b9
ck-gc.c:mark:102:CK_DEBUG:addr 0x7ffffe8: referenced middle = 0x1000b8
---------------------------------------------------------
checking for leaks:
ck-gc.c:sweep_leak:140:CK_DEBUG:checking pointer: 0x1000b8, start=0, middle=1
TRACE:	GC:Checked 1 blocks.
TRACE:		GC:SUCCESS: No leaks found!
----------------------------------------------------------
ck-gc.c:mark_all:207:stack has 30 words
ck-gc.c:mark:94:CK_DEBUG:checking pointer 0x1000b9
ck-gc.c:mark:102:CK_DEBUG:addr 0x7ffffe8: referenced middle = 0x1000b8
---------------------------------------------------------
checking for leaks:
ck-gc.c:sweep_leak:140:CK_DEBUG:checking pointer: 0x1000b8, start=0, middle=1
TRACE:ERROR:GC:MAYBE LEAK of 0x1000b8 (no pointer to the start)
TRACE:	nbytes=4
TRACE:	Block allocated at: tests/part2-test6.c:test1:7
TRACE:	GC:Checked 1 blocks.
TRACE:		GC:ERRORS: 0 errors, 1 maybe_errors
----------------------------------------------------------
TRACE:SUCCESS: detected a maybe leak!
DONE!!!

Saw done

bootloader: pi exited.  cleaning up
